File: packages/server/src/api/admin/templates.ts
import { Hono } from 'hono'
import { eq, and, ne } from 'drizzle-orm'
import { createTemplateSchema, updateTemplateSchema } from '@notify-hub/shared'
import { getDb, schema } from '../../db/index.js'

const templates = new Hono()

/**
 * GET /api/admin/templates
 */
templates.get('/', async (c) => {
  const db = getDb()
  const channelType = c.req.query('channelType')

  const all = channelType
    ? await db.select().from(schema.templates).where(eq(schema.templates.channelType, channelType))
    : await db.select().from(schema.templates)

  const result = all.map((t) => ({
    ...t,
    variables: t.variables ? JSON.parse(t.variables) : null,
  }))

  return c.json({ success: true, data: result })
})

/**
 * GET /api/admin/templates/:id
 */
templates.get('/:id', async (c) => {
  const db = getDb()
  const id = c.req.param('id')

  const [tpl] = await db
    .select()
    .from(schema.templates)
    .where(eq(schema.templates.id, id))
    .limit(1)

  if (!tpl) {
    return c.json({ success: false, error: 'Template not found' }, 404)
  }

  return c.json({
    success: true,
    data: { ...tpl, variables: tpl.variables ? JSON.parse(tpl.variables) : null },
  })
})

/**
 * POST /api/admin/templates
 */
templates.post('/', async (c) => {
  const body = await c.req.json()
  const parsed = createTemplateSchema.safeParse(body)

  if (!parsed.success) {
    return c.json({ success: false, error: parsed.error.issues[0].message }, 400)
  }

  const { name, channelType, subject, body: tplBody, variables } = parsed.data
  const db = getDb()

  // Check unique name
  const [existing] = await db
    .select({ id: schema.templates.id })
    .from(schema.templates)
    .where(eq(schema.templates.name, name))
    .limit(1)

  if (existing) {
    return c.json({ success: false, error: '模板名称已存在' }, 409)
  }

  const [result] = await db
    .insert(schema.templates)
    .values({
      name,
      channelType,
      subject: subject ?? null,
      body: tplBody,
      variables: variables ? JSON.stringify(variables) : null,
    })
    .returning()

  return c.json({ success: true, data: { id: result.id } }, 201)
})

/**
 * PUT /api/admin/templates/:id
 */
templates.put('/:id', async (c) => {
  const id = c.req.param('id')
  const body = await c.req.json()
  const parsed = updateTemplateSchema.safeParse(body)

  if (!parsed.success) {
    return c.json({ success: false, error: parsed.error.issues[0].message }, 400)
  }

  const db = getDb()
  const updates: Record<string, unknown> = {}

  if (parsed.data.name !== undefined) {
    // Check unique name (exclude self)
    const [dup] = await db
      .select({ id: schema.templates.id })
      .from(schema.templates)
      .where(and(eq(schema.templates.name, parsed.data.name), ne(schema.templates.id, id)))
      .limit(1)

    if (dup) {
      return c.json({ success: false, error: '模板名称已存在' }, 409)
    }
    updates.name = parsed.data.name
  }
  if (parsed.data.channelType !== undefined) updates.channelType = parsed.data.channelType
  if (parsed.data.subject !== undefined) updates.subject = parsed.data.subject
  if (parsed.data.body !== undefined) updates.body = parsed.data.body
  if (parsed.data.variables !== undefined) {
    updates.variables = parsed.data.variables ? JSON.stringify(parsed.data.variables) : null
  }

  await db
    .update(schema.templates)
    .set(updates)
    .where(eq(schema.templates.id, id))

  return c.json({ success: true })
})

/**
 * DELETE /api/admin/templates/:id
 */
templates.delete('/:id', async (c) => {
  const id = c.req.param('id')
  const db = getDb()

  await db.delete(schema.templates).where(eq(schema.templates.id, id))
  return c.json({ success: true })
})

export { templates }

File: crates/server/src/routes/templates.rs
use axum::extract::{Path, State};
use axum::routing::get;
use axum::{Json, Router};

use notifyhub_common::error::AppError;
use notifyhub_common::schemas::{CreateTemplateRequest, UpdateTemplateRequest};
use notifyhub_common::types::{ApiResponse, Template};
use notifyhub_common::constants::ChannelType;

use crate::auth::middleware::{AuthUser, require_admin};
use crate::AppState;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/admin/templates", get(list_templates).post(create_template))
        .route("/api/admin/templates/{id}", get(get_template).put(update_template).delete(delete_template))
}

async fn list_templates(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<Json<ApiResponse<Vec<Template>>>, AppError> {
    require_admin(&auth)?;

    let rows: Vec<TemplateRow> = sqlx::query_as("SELECT * FROM templates ORDER BY created_at DESC")
        .fetch_all(&state.pool)
        .await?;

    Ok(Json(ApiResponse::ok(rows.into_iter().map(Template::from).collect())))
}

async fn get_template(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<Template>>, AppError> {
    require_admin(&auth)?;

    let row: Option<TemplateRow> = sqlx::query_as("SELECT * FROM templates WHERE id = ?")
        .bind(&id)
        .fetch_optional(&state.pool)
        .await?;

    match row {
        Some(r) => Ok(Json(ApiResponse::ok(Template::from(r)))),
        None => Err(AppError::NotFound("template not found".into())),
    }
}

async fn create_template(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(req): Json<CreateTemplateRequest>,
) -> Result<Json<ApiResponse<Template>>, AppError> {
    require_admin(&auth)?;

    let id = uuid::Uuid::new_v4().to_string();
    let now = chrono::Utc::now().timestamp();
    let variables_json = req.variables.map(|v| serde_json::to_string(&v).unwrap_or_default());

    sqlx::query(
        "INSERT INTO templates (id, name, channel_type, subject, body, variables, created_at) VALUES (?, ?, ?, ?, ?, ?, ?)"
    )
    .bind(&id)
    .bind(&req.name)
    .bind(req.channel_type.as_str())
    .bind(&req.subject)
    .bind(&req.body)
    .bind(&variables_json)
    .bind(now)
    .execute(&state.pool)
    .await?;

    get_template(State(state), auth, Path(id)).await
}

async fn update_template(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(id): Path<String>,
    Json(req): Json<UpdateTemplateRequest>,
) -> Result<Json<ApiResponse<Template>>, AppError> {
    require_admin(&auth)?;

    let existing: Option<TemplateRow> = sqlx::query_as("SELECT * FROM templates WHERE id = ?")
        .bind(&id)
        .fetch_optional(&state.pool)
        .await?;
    existing.ok_or_else(|| AppError::NotFound("template not found".into()))?;

    if let Some(ref name) = req.name {
        sqlx::query("UPDATE templates SET name = ? WHERE id = ?").bind(name).bind(&id).execute(&state.pool).await?;
    }
    if let Some(ct) = req.channel_type {
        sqlx::query("UPDATE templates SET channel_type = ? WHERE id = ?").bind(ct.as_str()).bind(&id).execute(&state.pool).await?;
    }
    if let Some(ref subject) = req.subject {
        sqlx::query("UPDATE templates SET subject = ? WHERE id = ?").bind(subject).bind(&id).execute(&state.pool).await?;
    }
    if let Some(ref body) = req.body {
        sqlx::query("UPDATE templates SET body = ? WHERE id = ?").bind(body).bind(&id).execute(&state.pool).await?;
    }
    if let Some(ref variables) = req.variables {
        let v = serde_json::to_string(variables)?;
        sqlx::query("UPDATE templates SET variables = ? WHERE id = ?").bind(&v).bind(&id).execute(&state.pool).await?;
    }

    get_template(State(state), auth, Path(id)).await
}

async fn delete_template(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    require_admin(&auth)?;

    let result = sqlx::query("DELETE FROM templates WHERE id = ?")
        .bind(&id).execute(&state.pool).await?;

    if result.rows_affected() == 0 {
        return Err(AppError::NotFound("template not found".into()));
    }
    Ok(Json(ApiResponse::success()))
}

#[derive(sqlx::FromRow)]
struct TemplateRow {
    id: String,
    name: String,
    channel_type: String,
    subject: Option<String>,
    body: String,
    variables: Option<String>,
    created_at: i64,
}

impl From<TemplateRow> for Template {
    fn from(r: TemplateRow) -> Self {
        Template {
            id: r.id,
            name: r.name,
            channel_type: r.channel_type.parse().unwrap_or(ChannelType::Push),
            subject: r.subject,
            body: r.body,
            variables: r.variables.and_then(|s| serde_json::from_str(&s).ok()),
            created_at: r.created_at,
        }
    }
}
